### ISSUE TITLE
Python[23] missing CharStreams

### ISSUE THREAD

**Issue by kaby76** [antlr/antlr4#3106]
While adding a Python3 target to my universal driver generator for Antlr grammars (which works off the poms or just analyzes your .g4 and code and figures out what you want), I notice that CharStreams does not exist in the Python targets, missing "CharStreams.fromString()", "CharStreams.fromPath()", etc., which I use already for Java, CSharp, and JavaScript. According to the Java documentation, "[The CharStreams] class represents the primary interface for creating CharStreams from a variety of sources as of 4.7." I can work around this by using the "deprecated" interface, but it's unclear whether the Python targets are supposed to have this class.

**Comment by ericvergnaud** [comment]
missing in all runtimes except Java, C# and JavaScript
feel free to feel the gap

**Comment by kaby76** [comment]
> missing in all runtimes except Java, C# and JavaScript
> feel free to feel the gap

Eeks! I'll work around it for now.

**Comment by minghu6** [comment]
They use class InputStream instead of CharStreams.

**Comment by kaby76** [comment]
> They use class InputStream instead of CharStreams.
> 
> `from antlr4.InputStream import InputStream`

Yes, I know. I have been using InputStream in my code generator [Python3 template](https://github.com/kaby76/Domemtech.Trash/blob/main/trgen/templates/Python3/Program.py#L57) for CI testing [antlr/grammars-v4](https://github.com/antlr/grammars-v4) for several weeks now.

The API CharStreams has been around in Java since 2017--four years ago--and deprecates the ANTLRFileStream and ANTLRInputStream classes. The corresponding classes in the Python3 runtime are FileStream and InputStream. But, these classes are not deprecated.

Providing a consistent API across runtimes is desirable:
* It eases the burden of moving from one language to another.
* Bugs that exist in one runtime can be applied consistently across runtimes, saving time.

--Ken

**Comment by minghu6** [comment]
oh. I got it. @kaby76 I also met the same issue when moving target from Java to Python3.
And there are much difference between 2nd Antlr4 book said and source code did currently.

Maybe that's why self-doc code and up-to-date examples are just so essential to community driven project.

**Comment by kaby76** [comment]
@minghu6 Documentation is always the last thing to get updated, and understandably so. Documentation isn't going to be a good solution for these problems: How can a person who is not an expert in Antlr write a parser driver? How can we automate this for testing? Each version of Antlr has differences in the APIs, the language base classes and methods, and the toolchains. A template-driven generator is a better solution. --Ken

**Comment by minghu6** [comment]
en, What's the template-driven generator, what's used for? Does it means a EDSL on antlr?

**Comment by kaby76** [comment]
Sorry, I shouldn't invent a phrase for which there is already one. Trgen is a [template processor](https://en.wikipedia.org/wiki/Template_processor) for Antlr grammars. It outputs an Antlr parser program for any of the target languages. If you have further questions, please contact me directly. --Ken Domino
